src/Microsoft.AspNet.Hosting/Internal/HostingEngine.cs

```diff
@@ -71,6 +71,8 @@ public virtual IDisposable Start()
                 async features =>
                 {
                     var httpContext = contextFactory.CreateHttpContext(features);
+                    httpContext.Response.RegisterForDispose(httpContext);
```

Comment:
There's still and ordering problem. If this registers first it will run first. I think you'd have to register after `await application` to ensure you'd be last.